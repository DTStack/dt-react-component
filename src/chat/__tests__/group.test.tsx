import React from 'react';
import { act, cleanup, fireEvent, render } from '@testing-library/react';
import moment from 'moment';
import '@testing-library/jest-dom/extend-expect';

import Group from '../group';

function generateConversation() {
    const conversation = [
        {
            id: 'conversation_1',
            createdAt: 1736479532239,
            updatedAt: moment().subtract(1, 'day').toDate().getTime(),
            title: 'this is conversation 1',
            assistantId: 'assistant_1',
        },
        {
            id: 'conversation_2',
            createdAt: 1736479532239,
            updatedAt: moment().toDate().getTime(),
            title: 'this is conversation 2',
            assistantId: 'assistant_2',
        },
    ];
    return conversation;
}
jest.mock('../../ellipsisText', () => {
    return (props: any) => <div data-testid="ellipsisText">{props.value}</div>;
});
describe('Test Chat Group', () => {
    beforeEach(() => {
        cleanup();
    });

    it('Match snapshot', () => {
        const conversation = generateConversation();
        const onAdd = jest.fn();
        expect(render(<Group data={[]} />).asFragment()).toMatchSnapshot();
        expect(render(<Group data={[]} expand />).asFragment()).toMatchSnapshot('expand');
        expect(render(<Group data={[]} fullscreen />).asFragment()).toMatchSnapshot('fullscreen');
        expect(render(<Group data={[]} openFloat={false} />).asFragment()).toMatchSnapshot(
            'openFloat false'
        );
        expect(
            render(
                <Group data={[]} openFloat={false} dialogButtonProps={{ onClick: onAdd }} />
            ).asFragment()
        ).toMatchSnapshot('addNew');
        expect(render(<Group data={conversation} />).asFragment()).toMatchSnapshot('normal');
        expect(
            render(
                <Group data={conversation} listProps={{ selectId: conversation[0].id }} />
            ).asFragment()
        ).toMatchSnapshot('select');
    });
    it('Should fullscreen', () => {
        const { container } = render(<Group data={[]} fullscreen />);

        const ele = container.querySelector('.dtc-aigc-group');
        expect(ele).toBeInTheDocument();
        expect(ele?.className).toContain('dtc-aigc-group__fullscreen');
    });

    it('Should expand', () => {
        const { container } = render(<Group data={[]} expand />);

        const ele = container.querySelector('.dtc-aigc-group');
        expect(ele).toBeInTheDocument();
        expect(ele?.className).toContain('dtc-aigc-group__expand');
    });

    it('Should select item', () => {
        const conversation = generateConversation();
        const { container } = render(
            <Group data={conversation} listProps={{ selectId: conversation[0].id }} />
        );

        const ele = container.querySelector('.dtc-aigc-dialog-list-item');
        expect(ele).toBeInTheDocument();
        expect(ele?.className).toContain('dtc-aigc-dialog-list-item__selected');
    });

    it('Should group list title', () => {
        const { container } = render(<Group data={generateConversation()} />);

        const ele = container.querySelector('.dtc-aigc-group-list-item-title');
        expect(ele).toBeInTheDocument();
        expect(ele).toHaveTextContent('昨天');
    });

    it('Should support add new session', () => {
        const onAdd = jest.fn();
        const { container } = render(
            <Group data={[]} openFloat={false} dialogButtonProps={{ onClick: onAdd }} />
        );

        const btn = container.querySelector<HTMLDivElement>('.dtc__aigc__button');
        expect(onAdd).not.toBeCalled();
        expect(btn).not.toBeNull();

        act(() => {
            fireEvent.click(btn!);
        });
        expect(onAdd).toBeCalledWith(
            expect.objectContaining({
                type: 'click',
            })
        );
    });
    it('Should support select item', () => {
        const conversation = generateConversation();
        const onItemClick = jest.fn();
        const { container } = render(
            <Group data={conversation} listProps={{ onItemClick, selectId: conversation[0].id }} />
        );

        const nodeList = container.querySelectorAll<HTMLDivElement>('.dtc-aigc-dialog-list-item');
        const ele = nodeList?.item(nodeList?.length - 1);

        expect(onItemClick).not.toBeCalled();
        expect(ele).not.toBeNull();

        fireEvent.click(ele!);
        expect(onItemClick).toBeCalledWith(conversation[conversation.length - 1]);
    });

    test('Should render rename input when rename menu click and do rename', () => {
        const conversation = generateConversation();
        const onRename = jest.fn();
        const { container } = render(
            <Group
                data={conversation}
                listProps={{
                    selectId: conversation[0].id,
                    onRename,
                }}
            />
        );

        const icon = container.querySelectorAll('.ant-dropdown-trigger')[0];
        expect(icon).toBeInTheDocument();

        act(() => {
            fireEvent.click(icon);
        });

        const dropdownMenuItems = document.querySelectorAll(
            '.ant-dropdown:not(.ant-dropdown-hidden) .ant-dropdown-menu-item'
        );
        expect(dropdownMenuItems).toHaveLength(2);

        fireEvent.click(dropdownMenuItems[0]);

        const ele = container.querySelector('.dtc-aigc-dialog-list-item-input');
        expect(ele).toBeInTheDocument();
        expect(ele).toHaveAttribute('value', conversation[0].title);

        fireEvent.change(ele!, { target: { value: 'new_title' } });
        fireEvent.keyDown(ele!, { keyCode: 13 });
        expect(onRename).toBeCalledWith(conversation[0], 'new_title');
    });
    test('Should render delete button', () => {
        const conversation = generateConversation();
        const onDelete = jest.fn();
        const { container } = render(
            <Group
                data={conversation}
                listProps={{
                    selectId: conversation[0].id,
                    onDelete,
                }}
            />
        );

        const icon = container.querySelectorAll('.ant-dropdown-trigger')[0];
        expect(icon).toBeInTheDocument();

        act(() => {
            fireEvent.click(icon);
        });

        const dropdownMenuItems = document.querySelectorAll(
            '.ant-dropdown:not(.ant-dropdown-hidden) .ant-dropdown-menu-item'
        );
        expect(dropdownMenuItems).toHaveLength(2);

        fireEvent.click(dropdownMenuItems[1]);
        expect(onDelete).toBeCalledWith(conversation[0]);
    });
});
