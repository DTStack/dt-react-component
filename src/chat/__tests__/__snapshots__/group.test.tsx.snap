// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Chat Group Match snapshot 1`] = `
<DocumentFragment>
  <div
    class="dtc-aigc-group"
  >
    <div
      class="dtc-aigc-group-list dtc-aigc-group-list__float dtc-aigc-group-list__hide"
    >
      <div
        class="dtc-aigc-group-list-wrapper"
      >
        <div
          class="ant-empty dtc-empty dtc-aigc-group-list__empty"
        >
          <div
            class="ant-empty-image"
          >
            <img
              src="empty_project.png"
            />
          </div>
          <div
            class="ant-empty-description"
          >
            暂无对话
          </div>
        </div>
      </div>
    </div>
    <div
      class="dtc-aigc-group-content"
    />
  </div>
</DocumentFragment>
`;

exports[`Test Chat Group Match snapshot: addNew 1`] = `
<DocumentFragment>
  <div
    class="dtc-aigc-group"
  >
    <div
      class="dtc-aigc-group-list dtc-aigc-group-list__hide"
    >
      <button
        class="ant-btn ant-btn-default ant-btn-background-ghost dtc__aigc__button dtc__aigc__button--secondary"
        type="button"
      >
        <span
          class="dtc__icon dtc__aigc__button__text"
          style="font-size: 16px;"
        >
          <svg
            fill="none"
            height="1em"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M14.5 4C14.5 3.44772 14.0523 3 13.5 3H2.5C1.94772 3 1.5 3.44771 1.5 4V11.4189C1.5 11.9712 1.94771 12.4189 2.5 12.4189H3.56093C3.70743 12.4189 3.84656 12.4832 3.94156 12.5947L5.22452 14.1008C5.52397 14.4524 6.06694 14.4524 6.36639 14.1008L7.94881 12.2431C8.12787 12.0329 8.10262 11.7174 7.89241 11.5383C7.6822 11.3592 7.36662 11.3845 7.18756 11.5947L5.79545 13.2289L4.70281 11.9462C4.41781 11.6117 4.00043 11.4189 3.56093 11.4189H2.5V4H13.5V7.875C13.5 8.15114 13.7239 8.375 14 8.375C14.2761 8.375 14.5 8.15114 14.5 7.875V4ZM4.25 6C4.25 5.72386 4.47386 5.5 4.75 5.5H10C10.2761 5.5 10.5 5.72386 10.5 6C10.5 6.27614 10.2761 6.5 10 6.5H4.75C4.47386 6.5 4.25 6.27614 4.25 6ZM4.75 8.5C4.47386 8.5 4.25 8.72386 4.25 9C4.25 9.27614 4.47386 9.5 4.75 9.5H7.75C8.02614 9.5 8.25 9.27614 8.25 9C8.25 8.72386 8.02614 8.5 7.75 8.5H4.75Z"
              fill="currentColor"
              fill-rule="evenodd"
            />
            <path
              clip-rule="evenodd"
              d="M12.25 8.75C12.25 8.47386 12.0261 8.25 11.75 8.25C11.4739 8.25 11.25 8.47386 11.25 8.75V10.75H9.25C8.97386 10.75 8.75 10.9739 8.75 11.25C8.75 11.5261 8.97386 11.75 9.25 11.75H11.25V13.75C11.25 14.0261 11.4739 14.25 11.75 14.25C12.0261 14.25 12.25 14.0261 12.25 13.75V11.75H14.25C14.5261 11.75 14.75 11.5261 14.75 11.25C14.75 10.9739 14.5261 10.75 14.25 10.75H12.25V8.75Z"
              fill="currentColor"
              fill-rule="evenodd"
            />
            <defs>
              <lineargradient
                gradientUnits="userSpaceOnUse"
                id="paint0_linear_5309_15152"
                x1="1.5"
                x2="16.6164"
                y1="3"
                y2="9.2251"
              >
                <stop
                  stop-color="#00BAC6"
                />
                <stop
                  offset="0.5"
                  stop-color="#0067FF"
                />
                <stop
                  offset="1"
                  stop-color="#450FDE"
                />
              </lineargradient>
              <lineargradient
                gradientUnits="userSpaceOnUse"
                id="paint1_linear_5309_15152"
                x1="8.75"
                x2="15.9738"
                y1="8.25"
                y2="10.8506"
              >
                <stop
                  stop-color="#00BAC6"
                />
                <stop
                  offset="0.5"
                  stop-color="#0067FF"
                />
                <stop
                  offset="1"
                  stop-color="#450FDE"
                />
              </lineargradient>
            </defs>
          </svg>
        </span>
        <span
          class="dtc__aigc__button__text"
        >
          开启新对话
        </span>
      </button>
      <div
        class="dtc-aigc-group-list-wrapper"
      >
        <div
          class="ant-empty dtc-empty dtc-aigc-group-list__empty"
        >
          <div
            class="ant-empty-image"
          >
            <img
              src="empty_project.png"
            />
          </div>
          <div
            class="ant-empty-description"
          >
            暂无对话
          </div>
        </div>
      </div>
    </div>
    <div
      class="dtc-aigc-group-content"
    />
  </div>
</DocumentFragment>
`;

exports[`Test Chat Group Match snapshot: expand 1`] = `
<DocumentFragment>
  <div
    class="dtc-aigc-group dtc-aigc-group__expand"
  >
    <div
      class="dtc-aigc-group-list dtc-aigc-group-list__float"
    >
      <div
        class="dtc-aigc-group-list-wrapper"
      >
        <div
          class="ant-empty dtc-empty dtc-aigc-group-list__empty"
        >
          <div
            class="ant-empty-image"
          >
            <img
              src="empty_project.png"
            />
          </div>
          <div
            class="ant-empty-description"
          >
            暂无对话
          </div>
        </div>
      </div>
    </div>
    <div
      class="dtc-aigc-group-content"
    />
  </div>
</DocumentFragment>
`;

exports[`Test Chat Group Match snapshot: fullscreen 1`] = `
<DocumentFragment>
  <div
    class="dtc-aigc-group dtc-aigc-group__fullscreen"
  >
    <div
      class="dtc-aigc-group-list dtc-aigc-group-list__float dtc-aigc-group-list__hide"
    >
      <div
        class="dtc-aigc-group-list-wrapper"
      >
        <div
          class="ant-empty dtc-empty dtc-aigc-group-list__empty"
        >
          <div
            class="ant-empty-image"
          >
            <img
              src="empty_project.png"
            />
          </div>
          <div
            class="ant-empty-description"
          >
            暂无对话
          </div>
        </div>
      </div>
    </div>
    <div
      class="dtc-aigc-group-content"
    />
  </div>
</DocumentFragment>
`;

exports[`Test Chat Group Match snapshot: normal 1`] = `
<DocumentFragment>
  <div
    class="dtc-aigc-group"
  >
    <div
      class="dtc-aigc-group-list dtc-aigc-group-list__float dtc-aigc-group-list__hide"
    >
      <div
        class="dtc-aigc-group-list-wrapper"
      >
        <div
          class="dtc-aigc-group-list-item"
        >
          <div
            class="dtc-aigc-group-list-item-title"
          >
            昨天
          </div>
          <div
            class="dtc-aigc-dialog-list"
          >
            <div
              class="dtc-aigc-dialog-list-item"
            >
              <div
                data-testid="ellipsisText"
              >
                this is conversation 1
              </div>
              <span
                class="ant-dropdown-trigger"
                data-mock-icon="MoreOutlined"
              />
            </div>
          </div>
        </div>
        <div
          class="dtc-aigc-group-list-item"
        >
          <div
            class="dtc-aigc-group-list-item-title"
          >
            今天
          </div>
          <div
            class="dtc-aigc-dialog-list"
          >
            <div
              class="dtc-aigc-dialog-list-item"
            >
              <div
                data-testid="ellipsisText"
              >
                this is conversation 2
              </div>
              <span
                class="ant-dropdown-trigger"
                data-mock-icon="MoreOutlined"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="dtc-aigc-group-content"
    />
  </div>
</DocumentFragment>
`;

exports[`Test Chat Group Match snapshot: openFloat false 1`] = `
<DocumentFragment>
  <div
    class="dtc-aigc-group"
  >
    <div
      class="dtc-aigc-group-list dtc-aigc-group-list__hide"
    >
      <button
        class="ant-btn ant-btn-default ant-btn-background-ghost dtc__aigc__button dtc__aigc__button--secondary"
        type="button"
      >
        <span
          class="dtc__icon dtc__aigc__button__text"
          style="font-size: 16px;"
        >
          <svg
            fill="none"
            height="1em"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M14.5 4C14.5 3.44772 14.0523 3 13.5 3H2.5C1.94772 3 1.5 3.44771 1.5 4V11.4189C1.5 11.9712 1.94771 12.4189 2.5 12.4189H3.56093C3.70743 12.4189 3.84656 12.4832 3.94156 12.5947L5.22452 14.1008C5.52397 14.4524 6.06694 14.4524 6.36639 14.1008L7.94881 12.2431C8.12787 12.0329 8.10262 11.7174 7.89241 11.5383C7.6822 11.3592 7.36662 11.3845 7.18756 11.5947L5.79545 13.2289L4.70281 11.9462C4.41781 11.6117 4.00043 11.4189 3.56093 11.4189H2.5V4H13.5V7.875C13.5 8.15114 13.7239 8.375 14 8.375C14.2761 8.375 14.5 8.15114 14.5 7.875V4ZM4.25 6C4.25 5.72386 4.47386 5.5 4.75 5.5H10C10.2761 5.5 10.5 5.72386 10.5 6C10.5 6.27614 10.2761 6.5 10 6.5H4.75C4.47386 6.5 4.25 6.27614 4.25 6ZM4.75 8.5C4.47386 8.5 4.25 8.72386 4.25 9C4.25 9.27614 4.47386 9.5 4.75 9.5H7.75C8.02614 9.5 8.25 9.27614 8.25 9C8.25 8.72386 8.02614 8.5 7.75 8.5H4.75Z"
              fill="currentColor"
              fill-rule="evenodd"
            />
            <path
              clip-rule="evenodd"
              d="M12.25 8.75C12.25 8.47386 12.0261 8.25 11.75 8.25C11.4739 8.25 11.25 8.47386 11.25 8.75V10.75H9.25C8.97386 10.75 8.75 10.9739 8.75 11.25C8.75 11.5261 8.97386 11.75 9.25 11.75H11.25V13.75C11.25 14.0261 11.4739 14.25 11.75 14.25C12.0261 14.25 12.25 14.0261 12.25 13.75V11.75H14.25C14.5261 11.75 14.75 11.5261 14.75 11.25C14.75 10.9739 14.5261 10.75 14.25 10.75H12.25V8.75Z"
              fill="currentColor"
              fill-rule="evenodd"
            />
            <defs>
              <lineargradient
                gradientUnits="userSpaceOnUse"
                id="paint0_linear_5309_15152"
                x1="1.5"
                x2="16.6164"
                y1="3"
                y2="9.2251"
              >
                <stop
                  stop-color="#00BAC6"
                />
                <stop
                  offset="0.5"
                  stop-color="#0067FF"
                />
                <stop
                  offset="1"
                  stop-color="#450FDE"
                />
              </lineargradient>
              <lineargradient
                gradientUnits="userSpaceOnUse"
                id="paint1_linear_5309_15152"
                x1="8.75"
                x2="15.9738"
                y1="8.25"
                y2="10.8506"
              >
                <stop
                  stop-color="#00BAC6"
                />
                <stop
                  offset="0.5"
                  stop-color="#0067FF"
                />
                <stop
                  offset="1"
                  stop-color="#450FDE"
                />
              </lineargradient>
            </defs>
          </svg>
        </span>
        <span
          class="dtc__aigc__button__text"
        >
          开启新对话
        </span>
      </button>
      <div
        class="dtc-aigc-group-list-wrapper"
      >
        <div
          class="ant-empty dtc-empty dtc-aigc-group-list__empty"
        >
          <div
            class="ant-empty-image"
          >
            <img
              src="empty_project.png"
            />
          </div>
          <div
            class="ant-empty-description"
          >
            暂无对话
          </div>
        </div>
      </div>
    </div>
    <div
      class="dtc-aigc-group-content"
    />
  </div>
</DocumentFragment>
`;

exports[`Test Chat Group Match snapshot: select 1`] = `
<DocumentFragment>
  <div
    class="dtc-aigc-group"
  >
    <div
      class="dtc-aigc-group-list dtc-aigc-group-list__float dtc-aigc-group-list__hide"
    >
      <div
        class="dtc-aigc-group-list-wrapper"
      >
        <div
          class="dtc-aigc-group-list-item"
        >
          <div
            class="dtc-aigc-group-list-item-title"
          >
            昨天
          </div>
          <div
            class="dtc-aigc-dialog-list"
          >
            <div
              class="dtc-aigc-dialog-list-item dtc-aigc-dialog-list-item__selected"
            >
              <div
                data-testid="ellipsisText"
              >
                this is conversation 1
              </div>
              <span
                class="ant-dropdown-trigger"
                data-mock-icon="MoreOutlined"
              />
            </div>
          </div>
        </div>
        <div
          class="dtc-aigc-group-list-item"
        >
          <div
            class="dtc-aigc-group-list-item-title"
          >
            今天
          </div>
          <div
            class="dtc-aigc-dialog-list"
          >
            <div
              class="dtc-aigc-dialog-list-item"
            >
              <div
                data-testid="ellipsisText"
              >
                this is conversation 2
              </div>
              <span
                class="ant-dropdown-trigger"
                data-mock-icon="MoreOutlined"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="dtc-aigc-group-content"
    />
  </div>
</DocumentFragment>
`;
